### PR TITLE
Plugin stop mechanism

### PR DESCRIPTION
**Pull request by moonsky219**
The whole procedure works like this:

main.py
1) main.py will broadcast MobileInsight.Main.StopService when user click stop xxx plugin.
2) main.py waits for MobileInsight.Plugin.StopServiceAck (or a 5 seconds timer expired)
3) main.py stops plugin forcefully

plugin
1) waiting for MobileInsight.Main.StopService intent
2) finish any closure work
3) broadcasting MobileInsight.Plugin.StopServiceAck

To utilize this mechanism, plugin need to add following extra code:
1) define a callback function for doing closure work and sending StopServiceAck back to main.py
```python
def on_broadcast(context, intent):
    '''
    This plugin is going to be stopped, finish closure work
    '''
    IntentClass = autoclass("android.content.Intent")
    intent = IntentClass()
    action = 'MobileInsight.Plugin.StopServiceAck'
    intent.setAction(action)
    try:
        mi2app_utils.pyService.sendBroadcast(intent)
    except Exception as e:
        import traceback
        analyzer.log_error(str(traceback.format_exc()))
```
2) register this callback for receiving MobileInsight.Main.StopService at initial function
```python
br = BroadcastReceiver(
        on_broadcast, actions=['MobileInsight.Main.StopService'])
br.start()
```

### CODE REVIEW

File: app/plugins/NasAnalysis/main.mi2app [hunk 1]
```diff
@@ -3,10 +3,30 @@ import os
...
 
 # Import MobileInsight modules
-from mobile_insight.monitor import OnlineMonitor
+from android.broadcast import BroadcastReceiver
+from jnius import autoclass
 from mobile_insight.analyzer import LteNasAnalyzer, UmtsNasAnalyzer
+from mobile_insight.monitor import OnlineMonitor
 import mi2app_utils
 
+def on_broadcast(context, intent):
```
Comment (yuanjieli) [inline app/plugins/NasAnalysis/main.mi2app:12]: Would be great if we can eliminate plugin developer's efforts to manually add `on_broadcast()`. One possibility is to overload the `mobileinsight-mobile/app/service/main.py` and add `on_broadcast()` here. In this way, **any** plugin can be stopped.

**Comment by yuanjieli** [comment]
I would suggest to move on_broadcast() to
mobileinsight-mobile/app/service/main.py. In this way, the plugin
developers do not need to manually add this function. We can also prevent
any unexpected deadlocks due to imperfect plugins.

**Comment by moonsky219** [comment]
@yuanjieli 
But I think the plugin also need to do cleaning work itself inside the registered function (for example on_broadcast here).

**Comment by yuanjieli** [comment]
@moonsky219 I see. But what would happen if the plugin developer forgets to do so? My main concern is that `on_broadcast` is currently not mandatory. We should (1) make it mandatory **programmatically** since it is critical; (2) simplify the developer's efforts to overload this function.

**Comment by moonsky219** [comment]
Yes, I totally agree with you. My current work around is letting main.py to wait for maximum 5 seconds before stopping the plugin compulsively. A better implementation is still needed.

**Review by yuanjieli — DISMISSED** [review]

File: app/plugins/NasAnalysis/main.mi2app (hunk 1 above)
Comment (zwyuan) [reply app/plugins/NasAnalysis/main.mi2app:12]: should define basic functionalities in `analyzer`, and let the plugins to reload/override it.

**Review by yuanjieli — DISMISSED** [review]

**Comment by moonsky219** [comment]
I remove the code about checking orphan log in main.py and support it in LoggingAnalyzer of NetLogger

**Review by yuanjieli — APPROVED** [review]

**Pull request merged**